### ISSUE TITLE
Unreachable code and decorators that redirect

### ISSUE THREAD

**Issue by BrianPugh** [microsoft/pyright#3767]
**Describe the bug**
When a function containing an infinite loop is decorated by a decorator that redirects the code elsewhere, Pyright incorrectly believes the body of the function will execute.

While this might sound like nonsense code, in my actual usecase, i have a symmetric server/worker architecture where they both run the same code in different configurations. For the server, executing the (expensive) decorated function will actually send a command off to a remote worker to perform the actual execution.

**Expected behavior**
Pyright should detect that `foo` actually gets executed, and the final `print` statement is actually reachable.

**Screenshots or Code**
```
def foo():
    print("inside foo")


def decorator(f):
    return foo


@decorator
def bar():
    while True:
        print("inside bar")


bar()

print("unreachable")
```

when executed:

```
MacBook-Air:$ python test.py
inside foo
unreachable
```

![Screen Shot 2022-07-30 at 12 32 08 PM](https://user-images.githubusercontent.com/14318576/181962725-0a2403e4-67f7-43c6-b5f8-dabe949ff9b1.png)

**Comment by erictraut** [comment]
Yeah, this looks like a pretty atypical case. 

Pyright correctly infers the return type of the undecorated function `bar` as `NoReturn`.

When pyright encounters an unannotated decorator function, it assumes that the decorator does not change the signature of the functions it is decorating. We've found from experience that's the safest assumption.

If your decorator is not intended to preserve the signature of the functions it decorates, you should properly annotate it to indicate your intent.

```python
def decorator(f: Callable[..., Any]) -> Callable[[], None]:
    return foo
```

**Comment by BrianPugh** [comment]
Makes sense, thank you very much!
